### PR TITLE
Feature/prefabs

### PR DESCRIPTION
**Pull request by AlbertVVila**
## Prefabs

- A prefab is a template of a gameobject. All the properties are copied except for the transform component.

- You can create a new prefab by selecting a GameObject and then in the inspector mark *isPrefab*.
prefab

- As you can see, another checkbox will appear, *Sync to prefab*. This allows this prefab to be updated whenever the prefab is updated, by default it is already checked on.

- We can also see an *Update Prefab* button. This allows to update a given prefab. This will cause that all prefab instances with *Sync to prefab* checked on, will be updated.
prefab2

- When a Prefab is created, it is saved by default in the Assets folder, "Prefabs". In case it doesn't exists, it is created in the Assets base directory.

- In the Hierarchy a GameObject marked as prefab is colored in blue. We still can see the difference in color when this prefab is active or not.

- Prefab instances can be duplicated, deleted as any other gameobject. However nested prefabs are not allowed, so you cannot drag and drop prefab instances inside other prefabs.

prefab3

- Drag N Drop: In order to add a prefab to the current scene you don't need to use Add Scene. You can simply drag from the panel browser the desired prefab to the hierarchy.

prefab4

- Spawn Prefab: During runtime, it is useful to spawn objects like items for instance. Now in the module scene you have 2 methods *Spawn(...)*. In this methods you fill the name of the prefab and if you want to, the parent where the prefab is gonna be instantiated. The difference between them is that in the second method we can also set the position and rotation of the prefab.

- If a prefab is not found, we can still load the scene with the latest changes that the GameObject had.

- Even if you uncheck the isPrefab checkbox, the Prefab won't be deleted, so you can rebind it again when you check on.

## Other Features
- *FindByUID*: new method to find a gameobject by UID (Faster than *FindByName*)
- Modified *FindByName* parameter order to improve polymorphism 
- Fixed several memory leaks in *Resources* (*Textures*, *Mesh*, *Material*, *Model*...)

### CODE REVIEW

File: Source/ModuleFileSystem.h [hunk 1]
```diff
@@ -98,8 +102,9 @@ class ModuleFileSystem :
 	void ListFilesWithExtension(const char* dir, std::set<std::string>& files);											// Saves all files found on dir and subdirs on a set
 	bool CopyFromOutsideFS(const char* source, const char* destination) const;
 	bool Copy(const char* source, const char* destination, const char* file) const;
-	bool Move(const char * source, const char* file, const char* newFile) const;
```
Comment (Rubetoman) [inline Source/ModuleFileSystem.h]: Why did you delete the Move function?

File: Source/ModuleScene.cpp [hunk 2]
```diff
@@ -736,6 +803,43 @@ void ModuleScene::ResetQuadTree() //deprecated
...
+bool ModuleScene::PrefabWasUpdated(unsigned UID) const
+{
+	Resource* scene = App->resManager->GetByName(name.c_str(), TYPE::SCENE);
+	int prefabTime = App->fsystem->GetModTime(std::string(IMPORTED_PREFABS + std::to_string(UID) + PREFABEXTENSION).c_str());
+	int sceneTime = App->fsystem->GetModTime(std::string(IMPORTED_SCENES + std::to_string(scene->GetUID()) + SCENEEXTENSION).c_str());
+	App->resManager->DeleteResource(scene->GetUID());
+	return prefabTime > sceneTime;
+}
+
+unsigned ModuleScene::CreatePrefab(GameObject * go)
```
Comment (Rubetoman) [inline Source/ModuleScene.cpp]: Code style: GameObject* go

File: Source/Prefab.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+#include "Resource.h"
+
+#include <vector>
+
+class GameObject;
+class JSON;
+class JSON_value;
+
+class Prefab : public Resource
```
Comment (Rubetoman) [inline Source/Prefab.h]: If it is going to be a resource it should be named ResourcePrefab

File: Source/ComponentLight.cpp [hunk 4]
```diff
@@ -41,12 +41,21 @@ ComponentLight::ComponentLight(const ComponentLight& component) : Component(comp
...
+
 	pointSphere = Sphere(component.pointSphere);
 	App->scene->lights.push_back(this);
 
+	CalculateGuizmos();
 	produceShadows = component.produceShadows;
 }
 
+
+ComponentLight* ComponentLight::Clone() const
```
Comment (Rubetoman) [inline Source/ComponentLight.cpp]: These changes have nothing to do with prefabs.

File: Source/ComponentTransform.cpp [hunk 5]
```diff
@@ -373,7 +373,11 @@ void ComponentTransform::Paste()
 
 void ComponentTransform::Reset()
 {
-	position = math::float3(0.f, 0.f, 0.f);
-	eulerRotation = math::float3(0.f, 0.f, 0.f);
-	scale = math::float3(1.0f, 1.0f, 1.0f);
+	math::float3 position = math::float3::zero;
```
Comment (Rubetoman) [inline Source/ComponentTransform.cpp]: These changes have nothing to do with prefabs.

File: Source/ResourceMaterial.cpp [hunk 6]
```diff
@@ -203,6 +203,7 @@ void ResourceMaterial::SaveMetafile(const char* file) const
 	json->AddValue("Material", *meta);
 	filepath += METAEXT;
 	App->fsystem->Save(filepath.c_str(), json->ToString().c_str(), json->Size());
+	RELEASE(json);
```
Comment (Rubetoman) [inline Source/ResourceMaterial.cpp]: These changes have nothing to do with prefabs.

File: Source/ResourceMesh.cpp [hunk 7]
```diff
@@ -317,6 +319,8 @@ void ResourceMesh::SetMesh(const char* meshData)
 	ComputeBBox();
 	SetMeshBuffers();
 	SetBboxBuffers();
+
```
Comment (Rubetoman) [inline Source/ResourceMesh.cpp]: These changes have nothing to do with prefabs.

File: Source/ResourceTexture.cpp [hunk 8]
```diff
@@ -270,6 +270,7 @@ void ResourceTexture::SaveMetafile(const char* file) const
 	json->AddValue("Texture", *meta);
 	filepath += METAEXT;
 	App->fsystem->Save(filepath.c_str(), json->ToString().c_str(), json->Size());
+	RELEASE(json);
```
Comment (Rubetoman) [inline Source/ResourceTexture.cpp]: These changes have nothing to do with prefabs.

File: Source/ComponentLight.cpp (hunk 4 above)
Comment (AlbertVVila) [reply Source/ComponentLight.cpp]: I made a commit *Fix in ComponentLight and GameObject cloning*. Without this fix, prefabs where not created correctly on GameObjects with lights

File: Source/ComponentTransform.cpp (hunk 5 above)
Comment (AlbertVVila) [reply Source/ComponentTransform.cpp]: Commit *Fix bug in Spawn prefab + Moved Order Space Partitioning before Resou… *. Component Transform was not reset properly.

File: Source/ResourceMaterial.cpp (hunk 6 above)
Comment (AlbertVVila) [reply Source/ResourceMaterial.cpp]: Explained in the description of the feature:
*Fixed several memory leaks in Resources (Textures, Mesh, Material, Model...)*

File: Source/ResourceMesh.cpp (hunk 7 above)
Comment (AlbertVVila) [reply Source/ResourceMesh.cpp]: Explained in the description of the feature:
Fixed several memory leaks in Resources (Textures, Mesh, Material, Model...)

File: Source/ResourceTexture.cpp (hunk 8 above)
Comment (AlbertVVila) [reply Source/ResourceTexture.cpp]: Explained in the description of the feature:
*Fixed several memory leaks in Resources (Textures, Mesh, Material, Model...)*

File: Source/ModuleFileSystem.h (hunk 1 above)
Comment (AlbertVVila) [reply Source/ModuleFileSystem.h]: I changed it to a more powerful method Copy. Works like Move but can also change the name of file moved. This is very usefull to copy from assets to library changing name of file to UID.

File: Source/ModuleScene.cpp (hunk 2 above)
Comment (AlbertVVila) [reply Source/ModuleScene.cpp]: Done

File: Source/Prefab.h (hunk 3 above)
Comment (AlbertVVila) [reply Source/Prefab.h]: 👍

**Comment by AlbertVVila** [comment]
Removed unnecessary files 😃

**Comment by AlbertVVila** [comment]
In order to properly build scripts i've added a prebuild event to create a Scripts folder before compiling scripts. I deleted all DLLs in the repository and unnecessary metas and assets

File: Source/ModuleFileSystem.h (hunk 1 above)
Comment (Rubetoman) [reply Source/ModuleFileSystem.h]: Then it should be named MoveAndRename, it is very dangerous because the function move is being used by the RM and if you fail by putting a different name a lot of bugs may occur. I would prefer if you keep the Move function and create a new one that can do that. For the moment we have been using Move and then Rename for this, it is easier to understand.

File: Source/ResourceMaterial.cpp (hunk 6 above)
Comment (Rubetoman) [reply Source/ResourceMaterial.cpp]: Yeah, but is a completely different topic, it should be as a separate PR, if later a bug is found we can't know if it was caused by the prefabs or the memory leaks fix.

**Comment by AlbertVVila** [comment]
Prefabs are already ResourcePrefab, some scripts cpp have been changed because I had to change parameters of a method and those scripts used it. I think you cannot preview it because of the encoding of the file, I've opened a bug to solve this.

File: Source/ModuleFileSystem.h (hunk 1 above)
Comment (AlbertVVila) [reply Source/ModuleFileSystem.h]: Not really, the initial Move name was incorrect. The method in any case is moving a file, it is only copying it. The only change I made is to add the possibility to copy it with a diferent name. I already checked where you used it in case it didn't work. Have you checked if it behaves incorrectly?

File: Source/ResourceMaterial.cpp (hunk 6 above)
Comment (AlbertVVila) [reply Source/ResourceMaterial.cpp]: There is a commit only for memory leak fixes. So if there is a bug related to them, you can rewind in the repository to the commit before this one. It is not mixed with other changes. Anyway for further fixes it would be nice to behave the way you described.

**Comment by AlbertVVila** [comment]
Perfect, I'm gonna change it before merging.
